NetscapeBookmarkParser.php

```diff
@@ -0,0 +1,222 @@
+     *
+     * @param string $filename Bookmark file to parse
+     *
+     * @return array An associative array containing parsed links
+     */
+    public function parseFile($filename)
+    {
+        return $this->parseString(file_get_contents($filename));
+    }
+
+    /**
+     * Parses a string containing Netscape-formatted bookmarks
+     *
+     * Output format:
+     *
+     *     Array
+     *     (
+     *         [0] => Array
+     *             (
+     *                 [note]  => Some comments about this link
+     *                 [pub]   => 1
+     *                 [tags]  => a list of tags
+     *                 [time]  => 1459371397
+     *                 [title] => Some page
+     *                 [uri]   => http://domain.tld:5678/some-page.html
+     *             )
+     *         [1] => Array
+     *             (
+     *                 ...
+     *             )
+     *     )
+     *
+     * @param string $bookmarkString String containing Netscape bookmarks
+     *
+     * @return array An associative array containing parsed links
+     */
+    public function parseString($bookmarkString) {
+        $i = 0;
+        $next = false;
+        $currentTag = $this->defaultTag;
+
+        $lines = explode("\n", $this->sanitizeString($bookmarkString));
+
+        foreach ($lines as $line_no => $line) {
+            if (preg_match('/^<h\d(.*?)>(.*?)<\/h\d>/i', $line, $m1)) {
+                // a tag is matched: set its value as current
+                $currentTag = $m1[2];
+                continue;
+
+            } elseif (preg_match('/^<\/DL>/i', $line)) {
+                // <DL> matched: stop using tag
+                $currentTag = $this->defaultTag;
+            }
+
+            if (preg_match('/<a/i', $line, $m2)) {
+                if ($currentTag) {
+                    $this->items[$i]['tags'] = $currentTag;
+                }
+
+                if (preg_match('/href="(.*?)"/i', $line, $m3)) {
+                    $this->items[$i]['uri'] = $m3[1];
+                } else {
+                    $this->items[$i]['uri'] = '';
+                }
+
+                if (preg_match('/<a(.*?)>(.*?)<\/a>/i', $line, $m4)) {
+                    $this->items[$i]['title'] = $m4[2];
+                } else {
+                    $this->items[$i]['title'] = 'untitled';
+                }
+
+                if (preg_match('/note="(.*?)"<\/a>/i', $line, $m5)) {
+                     $this->items[$i]['note'] = $m5[1];
+                } elseif (preg_match('/<dd>(.*?)<\//i', $line, $m6)) {
+                     $this->items[$i]['note'] = str_replace('<br>', "\n", $m6[1]);
+                } else {
+                    $this->items[$i]['note'] = '';
+                }
+
+                if (preg_match('/(tags?|labels?|folders?)="(.*?)"/i', $line, $m7)) {
+                     $this->items[$i]['tags'] = strtr($m7[2], ',', ' ');
+                } else {
+                    $this->items[$i]['tags'] = $currentTag;
+                }
```

Comment:
If I ask for a default tag, it might be a good idea to append it either it finds existing tags or not?